Add angular support: Storybook for Angular

alterx opened the pull request:
Issue: #269 

## What I did
This is the first attempt at adding support for Angular based, mostly, on the vue changes.

What works:
- [x] Added angular sample code to the `examples` folder
- [x] Rendering stories and creating stories out of existing angular components
- [x] Supports components with separate templates/css as well as single-file components.
- Add-ons:
  - [x] Notes
  - [x] Actions
  - [x] Links
  - [x] Knobs
  - [x] HMR

## How to test

```sh
yarn bootstrap --reset --core
cd examples/angular-cli
yarn storybook
open http://localhost:9009
```

alterx committed:
Angular (2+) support (#269)

This is the first attempt at adding support for Angular based, mostly, on the vue changes.

What works:
- Added angular sample code to the `examples` folder
- Rendering stories and creating stories out of existing angular components
- For the time being, this only supports one-page angular components

What's missing:
- Addons support
- Support for components that have a separate template

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Removing @storybook/angular dep as it doesn't exist

Signed-off-by: Carlos Vega <clmvega@gmail.com>

shilman committed:
Update example deps

- `start-storybook` / `build-storybook` should be run out of
`node_modules/.bin`
- update lerna config to not publish angular-test example

alterx committed:
Updating README

Signed-off-by: Carlos Vega <clmvega@gmail.com>

shilman commented:
Awesome possum!! Super excited @alterx and welcome to Team Storybook!

ndelangen commented:
I rebased 3.2 with master, 
rebased 3.3 with 3.2
and rebased this branch with 3.3

ndelangen committed:
SYNC dependencies

alterx committed:
Clarifying strategy used to render dynamic components in angular

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Fixing issues with duplicate types introduced by recent changes

We had TypeScript errors related to conflicting @types included in the root @storybook repo and the angular-cli sample. Also, we had some issues with TS not recognizing `describe`, `it` and other global functions exposed by jasmine.

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Testing add-ons

- Links
- Actions
- Notes

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen committed:
Update README.md

alterx committed:
Enabling support for components with separate templates/css

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Update `getstorybook` CLI

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Adding clearer examples for Notes and Actions add-ons

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen commented on addons/knobs/README.md:
````diff
@@ -37,7 +37,7 @@ Now, write your stories with knobs.
 ```js
 import { storiesOf } from '@storybook/react';
-import { withKnobs, text, boolean, number } from '@storybook/addon-knobs';
+import { addonKnobs, text, boolean, number } from '@storybook/addon-knobs';
````
This is wrong now..

ndelangen commented on addons/knobs/README.md:
```diff
@@ -52,14 +52,15 @@ stories.add('with a button', () => (
+const options = {};
 // Knobs as dynamic variables.
-stories.add('as dynamic variables', () => {
+stories.add('as dynamic variables', addonKnobs(options)(() => {
```
Must have been 'reverted' when rebasing

alterx committed:
Removing unneded deps and updating deps versions

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Adding angular to the cli generator

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Laying the foundation to fix the `knobs` add-on for angular

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Adding else clausle for stories not sent by the `knobs` add-on

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen commented on examples/angular-cli/src/stories/index.ts:
```diff
@@ -0,0 +1,161 @@
+      }
+    };
+  })
+  // .add('All knobs', () => {
```
Do these not work currently? or is there another reason why these are commented-out?

ndelangen commented on addons/knobs/src/angular/knob-wrapper.component.ts:
```diff
@@ -0,0 +1,10 @@
+@Component({
+  selector: 'knobs-component',
+  template: `
+    <div>
```
An empty div? Is this uncompleted / in progress or is this the way it should be?

ndelangen reviewed (commented):
Overall looks good! Will test on my machine..

alterx commented:
This is a WIP. I just don't trust my Mac anymore, it's old and I better
commit what I have

On Wed, Aug 2, 2017, 3:53 AM Norbert de Langen <notifications@github.com>
wrote:

> *@ndelangen* commented on this pull request.
>
> Overall looks good! Will test on my machine..
> ------------------------------
>
> In examples/angular-cli/src/stories/index.ts
> <https://github.com/storybooks/storybook/pull/1474#discussion_r130832842>:
>
> > +
> +storiesOf('Addon Knobs', module)
> +  .addDecorator(withKnobs)
> +  .add('Simple', () => {
> +    const name = text('Name', 'John Doe');
> +    const age = number('Age', 44);
> +    const content = `I am ${name} and I'm ${age} years old.`;
> +
> +    return {
> +      component: SimpleKnobsComponent,
> +      props: {
> +        content
> +      }
> +    };
> +  })
> +  // .add('All knobs', () => {
>
> Do these not work currently? or is there another reason why these are
> commented-out?
> ------------------------------
>
> In addons/knobs/src/angular/knob-wrapper.component.ts
> <https://github.com/storybooks/storybook/pull/1474#discussion_r130833085>:
>
> > @@ -0,0 +1,10 @@
> +import { Component, Input, Output, EventEmitter } from '@angular/core';
> +
> +@Component({
> +  selector: 'knobs-component',
> +  template: `
> +    <div>
>
> An empty div? Is this uncompleted / in progress or is this the way it
> should be?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/storybooks/storybook/pull/1474#pullrequestreview-53749751>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABhRmFxfeegmTpDNsV9muUoIq9T_9mGMks5sUEcugaJpZM4OYyPd>
> .
>
-- 
Ing. Carlos Vega Rodríguez
*Web Developer / Software Engineer  | +506 89131416 *

alterx committed:
Updating angular version to latest beta and fixing errors regarding `@types/node` dependencies

Downgraded the version of `@types/node` used for the knobs addon. There's a well known issue with TS + lerna repositories (hoisted deps and symlinks). This fixes the issue for the time being.

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Removing ts lint errors for jasmine

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen committed:
FIX lerna.json

ndelangen committed:
REVERT changes on cra-kitchen-sink stories/index

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

ndelangen committed:
FIX

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

alterx committed:
Experimenting with knobs addon.

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Cleaning up angular app on each story load

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

alterx committed:
Knobs addon and Angular app refactor

- Knobs addon now works (it had a bug in which the knobs would get cleaned *after* the new story was loaded.
- No longer using decarator/classes for dynamic component creation, this will allow us to "inherit" from the components we get from the user.

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Merge branch '269-angular-support' of github.com:storybooks/storybook into 269-angular-support

alterx committed:
Updating package.json versions to match other apps

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

ndelangen committed:
trying to fix after yarn workspaces change

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

ndelangen committed:
ADD storybook bins for angular

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

ndelangen committed:
SYNC versions && FIX

alterx committed:
Merge branch '269-angular-support' of github.com:storybooks/storybook into 269-angular-support

alterx committed:
Adding changes removed by last commit

Signed-off-by: Carlos Vega <clmvega@gmail.com>

ndelangen committed:
Update README.md

ndelangen committed:
REMOVE angular knobs from old-style import

ndelangen committed:
UPDATE angular readme

ndelangen committed:
CLEANUP angular default addons.js

ndelangen committed:
CLEANUP angular client/index.js

ndelangen commented on app/angular/src/server/iframe.html.js:
```diff
@@ -0,0 +1,72 @@
+import url from 'url';
```
I think these 2 files have been changed into a new method for the other apps

ndelangen commented on app/angular/src/server/index.html.js:
```diff
@@ -0,0 +1,79 @@
+import url from 'url';
```
I think these 2 files have been changed into a new method for the other apps

ndelangen commented on app/angular/tslint.json:
```diff
@@ -0,0 +1,135 @@
+{
```
We could move this up, if there's more typescript in the codebase? Would that work?

alterx replied:
Well, this one's got some Angular specific stuff, such as `"node_modules/codelyzer"`

ndelangen commented on docs/pages/basics/introduction/index.md:
```diff
@@ -1,7 +1,8 @@
----
+* * *
```
BROKEN

ndelangen commented on examples/angular-cli/e2e/app.e2e-spec.ts:
```diff
@@ -0,0 +1,15 @@
+import { AppPage } from './app.po';
```
These are tests generated by the angular-cli right?

alterx replied:
Yeah, that's right

ndelangen committed:
CLEANUP && FIX unit test

alterx committed:
Allowing custom elements inside an angular component

ndelangen committed:
CHANGE webpack setup to be aligned with other apps

ndelangen committed:
Merge branch '269-angular-support' of github.com:storybooks/storybook into 269-angular-support

ndelangen committed:
Update index.md

alterx committed:
Fixing the styles example

alterx committed:
Fixing bug in which HRM reload would cause a duplicate component hence duplicate handlers

alterx committed:
Merge branch '269-angular-support' of github.com:storybooks/storybook into 269-angular-support

ndelangen committed:
ADD build-test and smoke-test for angular example

ndelangen committed:
FIX build-storybook for angular

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

ndelangen committed:
Update index.md

ndelangen committed:
FIX smoke tests for angular

ndelangen committed:
FIX angular detection for CLI

ndelangen committed:
ADD test for angular CLI

ndelangen commented:
There's a problem with the cli test @Hypnosphi :
```
 getstorybook - the simplest way to add a storybook to your project. 

 • Detecting project type. ✓

    There seems to be a storybook already available in this project.
    Apply following command to force:
    
   getstorybook -f 


 getstorybook - the simplest way to add a storybook to your project. 

 • Detecting project type. ✓
 • Adding storybook support to your "Angular" app(node:1038) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Package `@storybook/angular` doesn't exist
(node:1038) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

 getstorybook - the simplest way to add a storybook to your project. 

 • Detecting project type. ✓
 • Adding storybook support to your "Meteor" app. ✓
 • Preparing to install dependencies. ✓
```

Yet the test passes 🚨 !

I'll try and get test to work well, though I may need some help, since I don't really understand how it works.. Can we mock the npm calls here? Testing the full install is rather out of scope for testing the CLI? And potentially really flaky and slow?

Hypnosphi commented:
That’s weird, #1826 was supposed to fix that. Will take a look

Hypnosphi commented on lib/cli/bin/generate.js:
```diff
@@ -167,6 +167,13 @@ switch (projectType) {
+    // eslint-disable-next-line
+    require('../generators/ANGULAR')
+      .then(commandLog('Adding storybook support to your "Angular" app'))
+      .then(end);
```
You need to add `catch(handleError)` here

I’ll consider refactoring generate.js into async function with a single try/catch block, so that no promise rejection is left unhandled

alterx committed:
Debouncing render calls

- Angular's bootstrapping process is async and sometimes storybook triggers two re renders. In these cases, we reach an inconsistent state in which we can't properly dispose the old view and listeners get duplicated. With this debounce function we make sure we're accounting for those times.

shilman reviewed (changes requested):
See my typescript concerns above. TBH have not gone through much of the angular code yet.

alterx commented on addons/knobs/package.json:
```diff
@@ -29,14 +29,14 @@
   },
   "devDependencies": {
-    "@types/node": "^7.0.12",
+    "@types/node": "~6.0.60",
```
In that specific case the TS definition was already there. I downgraded it because typescript has issues with monorepos and hoisting  and wasn't being able to compile.
There are several github issues with no real fix for the problem (that I could make work) other than making the versions match.

alterx commented on addons/knobs/src/angular/helpers.ts:
```diff
@@ -0,0 +1,131 @@
+import {
```
Knobs also uses TS and has angular specific code inside the `angular` folder.

alterx replied:
I think we may be able to pull it off without TS but since it was already there I just used it

alterx committed:
Unused onClick handler in button example

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Updating to latest Angular 5 beta

alterx committed:
Actually, button component is used several times, adding onClick palceholder

Hypnosphi committed:
Merge remote-tracking branch 'upstream/release/3.3' into 269-angular-support

Hypnosphi committed:
Update CLI snapshot

Hypnosphi committed:
Skip smoke tests for angular-cli

Hypnosphi commented:
Seems like `start-storybook` doesn't work with angular v2.3.1 which is used in fixtures: https://circleci.com/gh/storybooks/storybook/11910
```
ERROR in /tmp/storybook/node_modules/@types/jasminewd2/index.d.ts
(18,13): error TS2428: All declarations of 'Matchers' must have identical type parameters.

ERROR in /tmp/storybook/node_modules/@types/jasminewd2/index.d.ts
(42,13): error TS2430: Interface 'ArrayLikeMatchers<T>' incorrectly extends interface 'Matchers<ArrayLike<T>>'.
  Types of property 'toBe' are incompatible.
    Type '(expected: any, expectationFailOutput?: any) => Promise<void>' is not assignable to type '{ (expected: any, expectationFailOutput?: any): boolean; (expected: any, expectationFailOutput?: ...'.
      Type 'Promise<void>' is not assignable to type 'boolean'.

ERROR in /tmp/storybook/node_modules/@types/jasminewd2/index.d.ts
(43,20): error TS2304: Cannot find name 'Expected'.

ERROR in /tmp/storybook/node_modules/@types/jasminewd2/index.d.ts
(44,23): error TS2304: Cannot find name 'Expected'.

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(26,41): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(27,39): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(152,24): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(302,15): error TS2428: All declarations of 'Matchers' must have identical type parameters.

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(333,14): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(429,24): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /tmp/storybook/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(442,30): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/dom_renderer.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'Renderer2'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/dom_renderer.d.ts
(8,21): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'RendererFactory2'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/dom_renderer.d.ts
(8,39): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'RendererType2'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/browser/server-transition.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/browser/server-transition.d.ts
(8,36): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/dom_tokens.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/directives/ng_style.d.ts
(8,48): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'Renderer2'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/directives/ng_for_of.d.ts
(8,36): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'NgIterable'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/directives/ng_for_of.d.ts
(8,87): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'TrackByFunction'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/directives/ng_class.d.ts
(8,65): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'Renderer2'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/i18n/localization.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/location/location_strategy.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/common/src/location/platform_location.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/events/hammer_gestures.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/events/event_manager.d.ts
(8,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/dom_tokens.d.ts
(1,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/dom/debug/by.d.ts
(8,24): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'Predicate'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser/src/browser.d.ts
(1,58): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/platform_providers.d.ts
(1,10): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/platform_core_dynamic.d.ts
(8,23): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(8,102): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(8,118): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(8,140): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'ɵConsole'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(9,115): error TS2305: Module '"/tmp/storybook/node_modules/@angular/compiler/index"' has no exported member 'CompileReflector'.

ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/platform-browser-dynamic.d.ts
(1,33): error TS2305: Module '"/tmp/storybook/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in ./stories/index.ts
Module build failed: error while parsing tsconfig.json
 @ ./.storybook/config.js 6:2-23
 @ multi /tmp/storybook/app/angular/dist/server/config/polyfills.js /tmp/storybook/app/angular/dist/server/config/globals.js (webpack)-hot-middleware/client.js?reload=true ./.storybook/config.js
```

We should find the minimal supported angular version, and check it in `detect.js`

NOTE FOR MYSELF: `start-storybook --smoke-test` should exit with code 1 on errors to prevent hanging builds

Hypnosphi commented:
I created a new branch for easier reproduction: https://github.com/storybooks/storybook/tree/269-angular-support%402.3.1

Basically, I just replaced the contents of `examples/angular-cli` with `lib/cli/test/snapshots/angular-cli`

alterx commented:
Hmm, this could be an issue with the types and/or the jasmine version. Angular should be backward compatible (I've only tested 4 and 5 tho, could be wrong)

Hypnosphi committed:
Merge branch 'release/3.3' into 269-angular-support

Hypnosphi committed:
Turn smoke tests back on for angular-cli fixture

Hypnosphi commented:
#1851 made it possible to turn smoke tests for CLI fixture back on, so that we don't forget to fix the error

ndelangen committed:
Merge branch 'release/3.3' into 269-angular-support

ndelangen commented:
@alterx Is this something you can help with:
```
ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@types/jasminewd2/index.d.ts
(18,13): error TS2428: All declarations of 'Matchers' must have identical type parameters.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@types/jasminewd2/index.d.ts
(42,13): error TS2430: Interface 'ArrayLikeMatchers<T>' incorrectly extends interface 'Matchers<ArrayLike<T>>'.
  Types of property 'toBe' are incompatible.
    Type '(expected: any, expectationFailOutput?: any) => Promise<void>' is not assignable to type '{ (expected: any, expectationFailOutput?: any): boolean; (expected: any, expectationFailOutput?: ...'.
      Type 'Promise<void>' is not assignable to type 'boolean'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@types/jasminewd2/index.d.ts
(43,20): error TS2304: Cannot find name 'Expected'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@types/jasminewd2/index.d.ts
(44,23): error TS2304: Cannot find name 'Expected'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(26,41): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(27,39): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(152,24): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(302,15): error TS2428: All declarations of 'Matchers' must have identical type parameters.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(333,14): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(429,24): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/lib/cli/test/run/angular-cli/node_modules/@types/jasmine/index.d.ts
(442,30): error TS2314: Generic type 'Matchers<T>' requires 1 type argument(s).

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/dom_renderer.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'Renderer2'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/dom_renderer.d.ts
(8,21): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'RendererFactory2'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/dom_renderer.d.ts
(8,39): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'RendererType2'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/browser/server-transition.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/browser/server-transition.d.ts
(8,36): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/dom_tokens.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/directives/ng_style.d.ts
(8,48): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'Renderer2'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/directives/ng_for_of.d.ts
(8,36): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'NgIterable'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/directives/ng_for_of.d.ts
(8,87): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'TrackByFunction'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/directives/ng_class.d.ts
(8,65): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'Renderer2'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/i18n/localization.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/location/location_strategy.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/common/src/location/platform_location.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/events/hammer_gestures.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/events/event_manager.d.ts
(8,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/dom_tokens.d.ts
(1,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/dom/debug/by.d.ts
(8,24): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'Predicate'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser/src/browser.d.ts
(1,58): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/platform_providers.d.ts
(1,10): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/platform_core_dynamic.d.ts
(8,23): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(8,102): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'InjectionToken'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(8,118): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(8,140): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'ɵConsole'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(9,115): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/compiler/index"' has no exported member 'CompileReflector'.

ERROR in /Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/platform-browser-dynamic/src/platform-browser-dynamic.d.ts
(1,33): error TS2305: Module '"/Users/dev/Projects/GitHub/storybook/reviewing/node_modules/@angular/core/index"' has no exported member 'StaticProvider'.

ERROR in ./stories/index.ts
Module build failed: error while parsing tsconfig.json
```

ndelangen commented:
I want to merge this into 3.3 so bad, so i can start the refactor

alterx commented:
Is this happening only with Angular 2? 
I could take a look (not sure if I can do it during this weekend tho) but it looks like an issue with the versions. 
I never really tested it with 2. V4 was already out when I started this feature.
How do I reproduce?

Hypnosphi commented:
> How do I reproduce? 

https://github.com/storybooks/storybook/tree/269-angular-support%402.3.1

```sh
yarn bootstrap --reset --core
cd examples/angular-cli
yarn storybook
```

robwormald commented:
(ng core team here) this is rad! please reach out if y'all need any help at all.

alterx commented:
Hey guys, I'm back. Been super busy at work but I'm going to look into that last issue :D
Thanks, @Hypnosphi

alterx committed:
Merge branch 'release/3.3' of github.com:storybooks/storybook into 269-angular-support

alterx committed:
Updating yarn.lock file

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx commented:
Ok guys, I spent a fair amount of time digging into these webpack errors and found several things:

1) Some of the issues are related to the fact the the Angular CLI, prior to 1.0 had a different structure (it didn't have a `tsconfig.json` in the base folder). This caused most of the errors as the `ts-loader` looks for this file on every parent directory, and it tried to transpile a lot of files that didn't even need to be transpiled. This first issue can be fixed by simply adding a `tsconfig.json` file to the  root of the `angular-cli` project.

2) We had some issues with the multiple versions of TypeScript, I'm not a yarn/lerna master but it looks like we had some conflicting versions that were causing `2.0.10` to get installed. This is a fairly old version that causes some issues. Updating the version to a more recent one fixed some of the errors.

3) Last but not least, there was some errors cause by a missing jasmine import and an old Protractor version.

After fixing these issues I was able to get it to compile using Angular 2.3.1 so we can say that it wasn't really an angular issue. It's just that a some of the modules angular uses are coupled to specific versions of TypeScript. I've committed those changes to the branch @Hypnosphi created just to demonstrate that we can fix it. With that being said I'd recommend to use the `@angular/cli@1.0.0` as our baseline (this installs angular@4.0.0) and just add something to let users know they can install it in previous versions of Angular but will need to play a bit with TypeScript and dependencies to make it work.

Thoughts on this?

ps. That branch was also using `addonNotes` instead of `withNotes`

ndelangen commented:
@alterx I agree, Sounds like you fixed it? Let's get this merged so @shilman can spend his weekend refactoring the common bits from the apps. I know he's looking forward to it! 😉

alterx committed:
Fixing protractor and typescript issues with cli tests

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx commented:
Guys, there's a problem with the CLI test, the `latest-version` package cannot find `@storybook/angular` (of course, it hasn't been published yet) 

```
 getstorybook - the simplest way to add a storybook to your project. 

 • Detecting project type. ✓
 • Adding storybook support to your "Angular" app
     Package `@storybook/angular` doesn't exist
npm info lifecycle @storybook/cli@3.3.0-alpha.2~test: Failed to exec test script
npm ERR! Test failed.  See above for more details.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Exited with code 1
```
It looks like this is the last issue, the build issues are gone 🎉 
@ndelangen @Hypnosphi

alterx committed:
Bumping @storybook/* versions

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Updating template and snapshot version

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Mocking @storybook/angular version for the time being

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx commented:
I've added more changes to fix these tests (changes include mocking the `@storybook/angular` version only for this release. Now, the issue happens when attempting to install the `@storybook/angular` package.

Hypnosphi commented:
That's weird, it should link a local version. I'll take a look

alterx commented:
It is weird, I even had to create this https://github.com/storybooks/storybook/pull/1474/commits/9becf40098978cd95714293814651dfda43fc38f commit to mock the version (it tries to fetch the version from the npm registry using https://www.npmjs.com/package/latest-version

Hypnosphi committed:
Add @storybook/cli to list of cli internal deps

Hypnosphi commented:
But it is `3.3.0-alpha.0`, according to its `package.json`

alterx commented:
Hmm, I thought I bumped them all up here https://github.com/storybooks/storybook/pull/1474/commits/64b7d3daff07c27f1a3c8f570e66d8ce043aee9d

Hypnosphi commented:
Well, the `version` field left the same (actually you're not supposed to do this by hand, so it's ok)

Hypnosphi committed:
Update CLI snapshot

alterx commented:
Oh, ok, good to know that :)

ndelangen commented:
We're good to go, I think.
@shilman can you update your review?

alterx committed:
Updating react version to 16

Signed-off-by: Carlos Vega <clmvega@gmail.com>

shilman committed:
Upgrade to react16

alterx committed:
Merge branch '269-angular-support' of github.com:storybooks/storybook into 269-angular-support

alterx committed:
Knobs addon: helpers shouldn't be a ts file

Signed-off-by: Carlos Vega <clmvega@gmail.com>

alterx committed:
Removing unused typescript deps

Signed-off-by: Carlos Vega <clmvega@gmail.com>
